File: 44_wildcard_matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        """
            https://leetcode.com/problems/wildcard-matching/
            Time Complexity - O(mn)
            Space Complexity - O(mn)
            'm' and 'n' is the length of input strings.
        """
        if s is None or p is None:
            return False
        s_len = len(s)
        p_len = len(p)
        dp = [[False for _ in range(p_len + 1)] for _ in range(s_len + 1)]

        # empty string matched with empty string
        dp[0][0] = True
        # char with empty string
        for r in range(1, s_len + 1):
            dp[r][0] = False
        # empty string with pattern string
        for c in range(1, p_len + 1):
            if p[c - 1] == '*':
                dp[0][c] = dp[0][c - 1]

        for r in range(1, s_len + 1):
            for c in range(1, p_len + 1):
                # if cur_char in both string match or if cur_pattern char is '?'
                # extend the string
                if (s[r - 1] == p[c - 1] or p[c - 1] == '?') and dp[r - 1][c - 1]:
                    dp[r][c] = True
                # if cur_pattern char is *
                # extend from prev string (taking * into account) or
                # ignore '*'
                elif p[c - 1] == '*' and (dp[r - 1][c] or dp[r][c - 1]):
                    dp[r][c] = True
        return dp[s_len][p_len]


if __name__ == '__main__':
    print(Solution().isMatch("aa", "a"))
    print(Solution().isMatch("adceb", "*a*b"))
    print(Solution().isMatch("aa", "*"))

File: 44-WildcardMatching.cpp
/* 
    Time Complexity                              :  dynamicProgramming - O(m*n)
                                                    twoPointers - O(max(m,n))
    Space Complexity                             :  O(1)
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  

class Solution {
public:
    bool isMatch(string s, string p) {
        // return dynamicProgramming(s, p);
        return twoPointers(s,p);
    }
    
    
    bool dynamicProgramming(string s, string p) {
        int n = s.size();
        int m = p.size();
        vector<vector<bool>> dp(m+1,vector<bool>(n+1,false));
        dp[0][0] = true;
        
        // pattern
        for(int i=1;i<=m;i++) {
            //string
            for(int j=0;j<=n;j++) {
                if(j == 0) {
                    dp[i][j] = p[i-1] == '*' ? dp[i-1][j] : false;
                } else {
                    if(s[j-1] == p[i-1] or p[i-1] == '?') {
                        dp[i][j] = dp[i-1][j-1];
                    } else if(p[i-1] == '*') {
                        dp[i][j] = (dp[i-1][j] or dp[i][j-1]);
                    } else if(s[j-1] != p[i-1]) {
                        dp[i][j] = false;
                    }  
                }
                
            }
        }
        
        return dp[m][n];
    }
    
    
    bool twoPointers(string s, string p) {
        int ps=-1, ss=0, pi=0, si=0, m=s.size(), n=p.size();
        
        while(si < m) {
            if(pi < n and s[si] == p[pi] or p[pi] == '?') {
                si++;
                pi++;
            } else if(pi < n and p[pi] == '*') {
                ps = pi;
                ss = si;
                pi++;
            } else if(ps == -1) {
                return false;
            } else {
                pi = ps;
                ss = ss + 1;
                si = ss;
            }
        }
        
        while(pi < n) {
            if(p[pi] != '*') return false;
            pi++;
        }
        
        return true;
    }
    
};